zulip/integrations/rocket.chat/rocket.chat-mirror

```diff
@@ -0,0 +1,123 @@
+                    cfg: Cfg) -> Callable[[Dict[str, Any]], None]:
+    zulip_cfg = cfg["zulip"]
+    rocket_cfg = cfg["rocket"]
+    site_without_http = zulip_cfg["site"].replace("https://", "").replace("http://", "")
+
+    def _zulip_to_rocket(msg: Dict[str, Any]) -> None:
+        """Zulip -> Matrix
+        """
+        isa_stream = msg["type"] == "stream"
+        not_from_bot = msg["sender_email"] != zulip_cfg["email"]
+        in_the_specified_stream = msg["display_recipient"] == zulip_cfg["stream"]
+        at_the_specified_subject = msg["subject"] == zulip_cfg["subject"]
+        if isa_stream and not_from_bot and in_the_specified_stream and at_the_specified_subject:
+            rocket_username = zulip_to_rocket_username(msg["sender_full_name"], site_without_http)
+            rocket_text = "{0}: {1}".format(rocket_username,
+                                            msg["content"])
+            pprint(rocket_client.chat_post_message(rocket_text, channel=rocket_cfg["channel_id"]).json())
+    return _zulip_to_rocket
+
+def rocket_listener(rocket_client: RocketChat, zulip_client: zulip.Client, cfg: Cfg):
+    interval = 2
+    rocket_cfg = cfg["rocket"]
+    while True:
+        now = datetime.datetime.utcnow()
+        oldest = now - datetime.timedelta(seconds=interval)
+        oldest_str = str(oldest)[:-3] + 'Z'
+        res = rocket_client.channels_history(rocket_cfg['channel_id'],
+                                             oldest=oldest_str).json()
+        rocket_to_zulip(zulip_client, cfg, res)
+        time.sleep(interval)
+
+if __name__ == '__main__':
+    signal.signal(signal.SIGINT, die)
+    logging.basicConfig(level=logging.WARNING)
+
+    # Get config for each clients
+    zulip_config = config["zulip"]
+    rocket_config = config["rocket"]
+
+    # Initiate clients
+    print("Starting rocketchat mirroring bot")
+
+    while True:
+        try:
+            zulip_client = zulip.Client(email=zulip_config["email"],
+                                        api_key=zulip_config["api_key"],
+                                        site=zulip_config["site"])
+
+            rocket_client = RocketChat(rocket_config['username'],
+                                       rocket_config['password'],
+                                       server_url=rocket_config['server_url'])
+
+            # A bidirectional mirror
+
+            # returns configured function handler
+            zulip_message_handler = zulip_to_rocket(rocket_client, config)
+            p1 = mp.Process(target=zulip_client.call_on_each_message,
+                            args=(zulip_message_handler,))
+
+            p2 = mp.Process(target=rocket_listener,
+                            args=(rocket_client, zulip_client, config))
+
+            print("Starting message handler on Zulip client")
+            p1.start()
+            print("Starting message handler on Rocket.Chat client")
+            p2.start()
+
+            p1.join()
```

Comment:
Is there a more idiomatic way to handle two forever-running subprocesses?

Reply:
Yes, there is only if one of the client has the feature. So far, only Matrix does: https://github.com/rht/python-zulip-api/blob/b94bf2a31325243f0c4e5402087f17fca77da79e/zulip/integrations/matrix/matrix_bridge.py#L94.

Reply:
If `zulip_client.call_on_each_message` can be made non-blocking, then no multiprocessing is necessary.

Reply:
I actually have no problem with these being multiprocess, since the two processes are essentially independent and don't require any coordination.  I'm more concerned about tearing them down, so it might make sense to use something like supervisord.

Reply:
Yeah, I'm concerned with the zombie processes even with `p.join()`. To remain lightweight, what about using multithreading like the Matrix client does[1]?

[1] https://github.com/matrix-org/matrix-python-sdk/blob/cb815ad96c60573f189ead02037f4d1207554717/matrix_client/client.py#L409-L414

Reply:
I actually prefer multiple processes here.